### PR TITLE
TLS metrics in `blackbox-exporter`

### PR DESCRIPTION
**Pull request by mimiteto**
**How to categorize this PR?**

/area monitoring
/kind enhancement

**What this PR does / why we need it**:
PR allows `probe_ssl_earliest_cert_expiry` to be kept for the `garden` Prometheus instance and adds the `kubernetes.default.svc.cluster.local` service to the target list.

**Which issue(s) this PR fixes**:
Provides a partial solution for #https://github.com/gardener/gardener/issues/10394

**Special notes for your reviewer**:

**Release note**:

```other operator
Introducing alerts for the expiration of the runtime API, dashboard, virtual API and gardener API CA certificates
```

### CODE REVIEW

File: pkg/component/observability/monitoring/blackboxexporter/blackboxexporter.go [hunk 1]
```diff
@@ -52,11 +52,15 @@ const (
 
 	volumeNameClusterAccess = "cluster-access"
 	// VolumeMountPathClusterAccess is the volume mount path to the cluster access credentials.
-	VolumeMountPathClusterAccess = "/var/run/secrets/blackbox_exporter/cluster-access"
+	VolumeMountPathClusterAccess = "/var/run/secrets/blackbox_exporter/" + volumeNameClusterAccess
```
Comment (rfranzke) [inline pkg/component/observability/monitoring/blackboxexporter/blackboxexporter.go]: It is conceptually wrong to use the `volumeName` constant here (even if the value is the same). These are two different things

File: pkg/component/observability/monitoring/blackboxexporter/blackboxexporter.go [hunk 2]
```diff
@@ -52,11 +52,15 @@ const (
...
 
 	volumeNameGardenerCA = "gardener-ca"
+	volumeNameRuntimeCA  = "runtime-ca"
 	// VolumeMountPathGardenerCA is the volume mount path to the gardener CA certificate bundle.
-	VolumeMountPathGardenerCA = "/var/run/secrets/blackbox_exporter/gardener-ca"
+	VolumeMountPathGardenerCA = "/var/run/secrets/blackbox_exporter/" + volumeNameGardenerCA
+	VolumeMountPathRuntimeCA  = "/var/run/secrets/blackbox_exporter/" + volumeNameRuntimeCA
```
Comment (rfranzke) [inline pkg/component/observability/monitoring/blackboxexporter/blackboxexporter.go]: same here

Comment (rfranzke) [inline pkg/component/observability/monitoring/blackboxexporter/blackboxexporter.go]: doc string missing

File: pkg/component/observability/monitoring/blackboxexporter/blackboxexporter.go [hunk 3]
```diff
@@ -52,11 +52,15 @@ const (
...
+	runtimeCAConfigMapName    = "kube-root-ca.crt"
+	RuntimeCAConfigMapKey     = "ca.crt"
```
Comment (rfranzke) [inline pkg/component/observability/monitoring/blackboxexporter/blackboxexporter.go]: doc string missing

File: pkg/component/observability/monitoring/blackboxexporter/blackboxexporter_test.go [hunk 4]
```diff
@@ -97,6 +98,7 @@ var _ = Describe("BlackboxExporter", func() {
 			},
 		}
 
+		format.MaxLength = 100000
```
Comment (rfranzke) [inline pkg/component/observability/monitoring/blackboxexporter/blackboxexporter_test.go]: remove this

File: pkg/component/observability/monitoring/blackboxexporter/garden/config.go [hunk 5]
```diff
@@ -44,16 +45,21 @@ func Config(isDashboardCertificateIssuedByGardener, isGardenerDiscoveryServerEna
 		httpKubeAPIServerRootCAsModule    = defaultModuleConfig()
 		httpGardenerDashboardModule       = defaultModuleConfig()
 		httpGardenerDiscoveryServerModule = defaultModuleConfig()
+		httpRuntimeAPIServerModule        = defaultModuleConfig()
 
 		pathGardenerAPIServerCABundle = blackboxexporter.VolumeMountPathGardenerCA + "/" + secretsutils.DataKeyCertificateBundle
 		pathKubeAPIServerCABundle     = blackboxexporter.VolumeMountPathClusterAccess + "/" + secretsutils.DataKeyCertificateBundle
+		pathRuntimeAPIServerCABundle  = blackboxexporter.VolumeMountPathRuntimeCA + "/" + blackboxexporter.RuntimeCAConfigMapKey
 		pathToken                     = blackboxexporter.VolumeMountPathClusterAccess + "/" + resourcesv1alpha1.DataKeyToken
+		pathToRuntimeToken            = blackboxexporter.VolumeMountPathRuntimeCA + "/" + resourcesv1alpha1.DataKeyToken
```
Comment (rfranzke) [inline pkg/component/observability/monitoring/blackboxexporter/garden/config.go]:
```suggestion
		pathRuntimeToken            = blackboxexporter.VolumeMountPathRuntimeCA + "/" + resourcesv1alpha1.DataKeyToken
```

File: pkg/component/observability/monitoring/blackboxexporter/garden/scrapeconfig.go [hunk 6]
```diff
@@ -14,8 +14,16 @@ import (
...
+func ScrapeConfig(namespace string, kubeAPIServerTargets []monitoringv1alpha1.Target, gardenerDashboardTarget monitoringv1alpha1.Target, isGarden bool) []*monitoringv1alpha1.ScrapeConfig {
+	defaultKeepMetrics := []string{
+		"probe_success",
+		"probe_http_status_code",
+		"probe_http_duration_seconds",
+	}
+	additionalKeepTLSMetrics := []string{
+		"probe_ssl_earliest_cert_expiry",
+	}
+	genScrapeConfig := func(name, module string, targets []monitoringv1alpha1.Target, metrics []string) *monitoringv1alpha1.ScrapeConfig {
```
Comment (rfranzke) [inline pkg/component/observability/monitoring/blackboxexporter/garden/scrapeconfig.go]:
```suggestion
	generateScrapeConfig := func(name, module string, targets []monitoringv1alpha1.Target, metrics []string) *monitoringv1alpha1.ScrapeConfig {
```

Comment (rfranzke) [inline pkg/component/observability/monitoring/blackboxexporter/garden/scrapeconfig.go]: Use `var ()` block

File: pkg/component/observability/monitoring/blackboxexporter/garden/scrapeconfig.go [hunk 7]
```diff
@@ -55,19 +63,21 @@ func ScrapeConfig(namespace string, kubeAPIServerTargets []monitoringv1alpha1.Ta
...
-					"probe_http_duration_seconds",
-				),
+				MetricRelabelConfigs: monitoringutils.StandardMetricRelabelConfig(metrics...),
 			},
 		}
 	}
+	defaultScrapeConfig := func(name, module string, targets []monitoringv1alpha1.Target) *monitoringv1alpha1.ScrapeConfig {
+		return genScrapeConfig(name, module, targets, defaultKeepMetrics)
+	}
+	tlsScrapeConfig := func(name, module string, targets []monitoringv1alpha1.Target) *monitoringv1alpha1.ScrapeConfig {
```
Comment (rfranzke) [inline pkg/component/observability/monitoring/blackboxexporter/garden/scrapeconfig.go]: `var ()` block

**Review by rfranzke — CHANGES_REQUESTED** [review]
/cc @istvanballok @rickardsjp @vicwicker @chrkl 

Generally, I have doubts whether this is a  good approach to alert about  the CA certificate expiry - what about all other CAs/certs that are generated in the cluster (which maybe cannot be used so easily in `blackbox-exporter`)?

File: pkg/component/observability/monitoring/prometheus/garden/prometheusrules.go [hunk 8]
```diff
@@ -196,6 +196,18 @@ func gardenPrometheusRule(isGardenerDiscoveryServerEnabled bool) *monitoringv1.P
 					"last hour. The pod is running on the garden cluster.",
 			},
 		},
+		{
+			Alert:  "RuntimeCACertificateAboutToExpire",
+			Expr:   intstr.FromString("probe_ssl_earliest_cert_expiry - time() < 86400 * 30"),
+			For:    ptr.To(monitoringv1.Duration("10m")),
```
Comment (vicwicker) [inline pkg/component/observability/monitoring/prometheus/garden/prometheusrules.go]: Maybe add a comment that this is supposed to be 30 days or break it down such as `24 * 3600 * 30`.

The for is optional and I think there is no need to wait 10 minutes for the alert to stay up before sending it. If a certificate is going to expire within X time then it's still going to expire 10 minutes later.

**Review by vicwicker — COMMENTED** [review]
Thanks for addressing the previous review comments I sent you offline. I reviewed again and only found one more thing.

**Pull request closed without merging**